resources/assets/js/components/H2Components/Dialog.tsx

```diff
@@ -123,39 +121,124 @@ const Trigger: React.FunctionComponent<TriggerProps> = (props) => {
-  Header: React.FunctionComponent<GeneralProps>;
-  Title: React.FunctionComponent<GeneralProps>;
+  Actions: React.FunctionComponent<ActionsProps>;
+  ActionBtn: React.FunctionComponent<ActionBtnProps>;
+  Content: React.FunctionComponent<ContentProps>;
+  Header: React.FunctionComponent<HeaderProps>;
+  Title: React.FunctionComponent<TitleProps>;
   Trigger: React.FunctionComponent<TriggerProps>;
 }
 
 const Dialog: React.FunctionComponent<DialogContext> & DialogComposition = (
   props,
 ) => {
-  const { id, className, children, ...rest } = props;
-  const ref = React.useRef(null);
+  const { id, isVisible, className, children, closeDialog, ...rest } = props;
+  // Focus the first focusable element when the modal becomes visible.
+  const dialogRef = React.useRef<HTMLDivElement | null>(null);
+  React.useEffect(() => {
+    if (isVisible && dialogRef.current) {
+      const focusableModalElements = getTabList(dialogRef.current);
+      if (focusableModalElements.length > 0) {
+        const firstElement = focusableModalElements[0] as HTMLElement;
+        firstElement.focus();
+      }
+    }
+  }, [isVisible]);
+
+  const handleKeyUp = React.useCallback(
+    (event) => {
+      switch (event.key) {
+        case "Escape":
+          closeDialog();
+          focusOnElement(`[data-h2-dialog-trigger=${id}]`);
+          break;
+        case "Tab":
+          if (dialogRef && dialogRef.current) {
+            const focusableDialogElements = getTabList(dialogRef.current);
+
+            if (focusableDialogElements.length === 0) {
+              event.preventDefault(); // TODO: should this throw an error?
+              return;
+            }
+
+            const firstElement = focusableDialogElements[0] as HTMLElement;
+            const lastElement = focusableDialogElements[
+              focusableDialogElements.length - 1
+            ] as HTMLElement;
+
+            if (focusableDialogElements.length === 1) {
+              // This check to avoid strange behavior if firstElement == lastElement.
+              firstElement.focus();
+              event.preventDefault();
+              return;
+            }
+
+            const focusableDialogElementsArray = Array.from(
+              focusableDialogElements,
+            );
+
+            if (
+              document.activeElement &&
+              !focusableDialogElementsArray.includes(
+                document.activeElement as HTMLElement,
+              )
+            ) {
+              firstElement.focus();
+              event.preventDefault();
+              return;
+            }
+
+            if (!event.shiftKey && document.activeElement === lastElement) {
+              firstElement.focus();
+              event.preventDefault();
+              return;
+            }
+
+            if (event.shiftKey && document.activeElement === firstElement) {
+              lastElement.focus();
+              event.preventDefault();
+            }
```

Comment:
Can/should we use the helper functions you wrote here?

Reply:
I don't really see any improvements using the helper methods here, what do you think?

```js
if (!event.shiftKey) {
  focusPreviousItem(tabableElements);
  event.preventDefault();
  return;
}
```

Reply:
Yeah, you're right. Your current solution allows the default behaviour to happen in most cases, which is probably for the best.